### PR TITLE
[lldb] Use Py_InitializeFromConfig with Python >= 3.8 (NFC)

### LINKED ISSUE #113475: [lldb] Fix deprecation warning because of Py_SetPythonHome
Currently building lldb with python support causes a deprecation warning in the build. This is caused by the use of Py_SetPythonHome() in [ScriptInterpreterPython.cpp:166. ](https://github.com/llvm/llvm-project/blob/a4ace3de1b390451ce5e41bce74b90dcd992ab61/lldb/source/Plugins/ScriptInterpreter/Python/ScriptInterpreterPython.cpp#L145) Which is deprecated since 3.11. Python is moving to using a different init mechanism, using the PyConfig struct instead, and relying on setting a config string within the struct. 
Relevant info: 
https://docs.python.org/3/c-api/init.html#c.Py_SetPythonHome
https://docs.python.org/3/c-api/init_config.html#init-config
https://docs.python.org/3/c-api/init_config.html#c.PyConfig
This is a good first issue.

### PR DESCRIPTION
**Pull request by JDevlieghere**
This fixes the deprecation warning for Py_SetPythonHome, which was deprecated in Python 3.11. With this patch, when building against Python 3.8 or later, we now use Py_InitializeFromConfig instead.

Fixes #113475

### CODE REVIEW

**Comment by JDevlieghere** [comment]
CC @R-Goc

**Review by medismailben — APPROVED** [review]
LGTM

**Comment by R-Goc** [comment]
Looks good.

File: lldb/source/Plugins/ScriptInterpreter/Python/ScriptInterpreterPython.cpp [hunk 1]
```diff
@@ -117,15 +148,22 @@ struct InitializePythonRAII {
...
+    config.install_signal_handlers = 0;
+    Py_InitializeFromConfig(&config);
+    PyConfig_Clear(&config);
+    InitializeThreadsPrivate();
+#else
 // Python < 3.2 and Python >= 3.2 reversed the ordering requirements for
 // calling `Py_Initialize` and `PyEval_InitThreads`.  < 3.2 requires that you
 // call `PyEval_InitThreads` first, and >= 3.2 requires that you call it last.
-#if (PY_MAJOR_VERSION == 3 && PY_MINOR_VERSION >= 2) || (PY_MAJOR_VERSION > 3)
+#if (PY_MAJOR_VERSION == 3 && PY_MINOR_VERSION >= 2)
```
Comment (DavidSpickett) [inline lldb/source/Plugins/ScriptInterpreter/Python/ScriptInterpreterPython.cpp:160]: Doesn't this now do the wrong thing for a hypothetical 4.x? I thought that's what the `|| (PY_MAJOR_VERSION > 3)` was there for.

File: lldb/source/Plugins/ScriptInterpreter/Python/ScriptInterpreterPython.cpp [hunk 2]
```diff
@@ -92,7 +92,38 @@ namespace {
...
+          return nullptr;
+        spec.GetPath(path);
+        llvm::sys::path::append(path, lldb_python_home);
+        absolute_python_home = path.c_str();
+      }
+      size_t size = 0;
+      return Py_DecodeLocale(absolute_python_home, &size);
+    }();
+    if (g_python_home != nullptr) {
+#if (PY_MAJOR_VERSION == 3 && PY_MINOR_VERSION >= 8)
```
Comment (DavidSpickett) [inline lldb/source/Plugins/ScriptInterpreter/Python/ScriptInterpreterPython.cpp:120]: Same here, `|| (PY_MAJOR_VERSION > 3)`.

File: lldb/source/Plugins/ScriptInterpreter/Python/ScriptInterpreterPython.cpp (hunk 1 above)
Comment (labath) [reply lldb/source/Plugins/ScriptInterpreter/Python/ScriptInterpreterPython.cpp:160]: [#if PY_VERSION_HEX>=0x03020000](https://github.com/python/cpython/blob/dcad8fecbd5c8f1542f7a46e0c7bb7e1f27ab115/Include/patchlevel.h#L31) ?

Comment (JDevlieghere) [reply lldb/source/Plugins/ScriptInterpreter/Python/ScriptInterpreterPython.cpp:160]: @DavidSpickett The hypothetical Python 4 case is covered above and goes down the ` Py_InitializeFromConfig` path. 

@labath Happy to use that but everything else in this file uses the same check I have here, so I'll simplify the defines in a separate patch/PR.

File: lldb/source/Plugins/ScriptInterpreter/Python/ScriptInterpreterPython.cpp (hunk 2 above)
Comment (JDevlieghere) [reply lldb/source/Plugins/ScriptInterpreter/Python/ScriptInterpreterPython.cpp:120]: Yes, here it's actually missing.

File: lldb/source/Plugins/ScriptInterpreter/Python/ScriptInterpreterPython.cpp (hunk 1 above)
Comment (DavidSpickett) [reply lldb/source/Plugins/ScriptInterpreter/Python/ScriptInterpreterPython.cpp:160]: Yes you're right I was struggling to read the nested ifdefs. This 3.2 code will be removed once the 20 branch happens so it's fine as is.

**Review by DavidSpickett — APPROVED** [review]
LGTM

**Pull request merged**